pandas/core/indexing.py

```diff
@@ -859,15 +860,22 @@ def _convert_for_reindex(self, key, axis=0):
             return labels[key]
         else:
             if isinstance(key, Index):
-                # want Index objects to pass through untouched
-                keyarr = key
+                if isinstance(labels, ABCUInt64Index) and key.is_integer():
+                    keyarr = key.astype(np.uint64)
+                else:
+                    keyarr = key
             else:
                 # asarray can be unsafe, NumPy strings are weird
                 keyarr = _asarray_tuplesafe(key)
 
-            if is_integer_dtype(keyarr) and not labels.is_integer():
-                keyarr = _ensure_platform_int(keyarr)
-                return labels.take(keyarr)
+            if is_integer_dtype(keyarr):
```

Comment:
same here, if there are too many special cases this makes this code even more impenetrable.